### PR TITLE
Standalone queryoptions package polymorphism

### PR DESCRIPTION
**Pull request by Nick-Lucas**
* Adds polymorphism tests where a shared router type can be used to create shared components
* Adds queryKey and queryFilter factories with tests
* Adds InferInput and InferOutput helpers which can be used against Query & Mutation Options factories to get their types

image

## 🎯 Changes

What changes are made in this PR? Is it a feature or a bug fix?

## ✅ Checklist

- [ ] I have followed the steps listed in the [Contributing guide](https://github.com/trpc/trpc/blob/main/CONTRIBUTING.md).
- [ ] If necessary, I have added documentation related to the changes made.
- [ ] I have added or updated the tests related to the changes made.

### CODE REVIEW

File: packages/tanstack-react-query/src/internals/createOptionsProxy.ts [hunk 1]
```diff
@@ -46,9 +84,33 @@ export interface DecorateQueryProcedure<TDef extends ResolverDef> {
...
 export interface DecorateMutationProcedure<TDef extends ResolverDef> {
+  /**
+   * @internal prefer using InferInput to access input type
+   */
+  '~input': TDef['input'];
+
+  /**
+   * @internal prefer using InferOutput to access input type
+   */
+  '~output': TDef['output'];
```
Comment (KATT) [inline packages/tanstack-react-query/src/internals/createOptionsProxy.ts:112]: nit: we usually use `$` to signify types

Comment (KATT) [reply packages/tanstack-react-query/src/internals/createOptionsProxy.ts:112]: could maybe be

```ts
$types: {
  input: x
  output: y
}
```
to align with e.g. our root obj

Comment (juliusmarminge) [reply packages/tanstack-react-query/src/internals/createOptionsProxy.ts:112]: i think the promise was that ~ gets sorted last in auto-complete, but I can't repro that 🤔 

CleanShot 2024-11-17 at 22 20 45

Comment (Nick-Lucas) [reply packages/tanstack-react-query/src/internals/createOptionsProxy.ts]: Pushed this as the probable outcome but happy to discuss further

image

Comment (Nick-Lucas) [reply packages/tanstack-react-query/src/internals/createOptionsProxy.ts]: @juliusmarminge I think the sorting trick doesn't work in object autocomplete but it does for autocompleting chaining, see:

image

Comment (juliusmarminge) [reply packages/tanstack-react-query/src/internals/createOptionsProxy.ts]: ah okay

**Review by juliusmarminge — APPROVED** [review]

**Pull request merged**